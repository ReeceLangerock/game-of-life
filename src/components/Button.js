import React from "react";
import styled from "styled-components";

const StyledButton = styled.button`
  background-color: ${props => (props.active ? "#80CF7C" : "white")};
  color: ${props => (props.active ? "white" : "black")};
  padding: 4px 11px;
  margin: 5px;
  cursor: pointer;
  border: none;
  box-shadow: 0 4px 8px rgba(105, 114, 115, 0.16), 0 3px 6px rgba(105, 114, 115, 0.23);
  transition: 0.5s ease-in-out;
  outline: none;
  &:hover {
    transition: 0.5s ease-in-out;
    box-shadow: 0 9px 18px rgba(71, 115, 69, 0.29), 0 5px 5px rgba(71, 115, 69, 0.33);
  }
`;

export class Button extends React.Component {
  render() {
    return (
      <StyledButton active={this.props.active} onClick={this.props.onClick}>
        {this.props.buttonText}
      </StyledButton>
    );
  }
}
export default Button;
