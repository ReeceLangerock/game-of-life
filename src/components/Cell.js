import React from "react";
import styled, { injectGlobal } from "styled-components";

const StyledCell = styled.div`
  width: ${props => 500 / props.size - 1}px;
  height: ${props => 500 / props.size - 1}px;

  // width: 6px;
  // height: 6px;
  border-top: 1px solid;
  border-left: 1px solid;
  border-collapse: collapse;
  background: ${props => (props.status ? "green" : "white")};
`;

export class Cell extends React.Component {

  shouldComponentUpdate(nextProps) {
    return nextProps.status !== this.props.status
}
  render() {
    return <StyledCell size={this.props.size} status={this.props.status} onClick={this.props.onClick} />;
  }
}
export default Cell;
