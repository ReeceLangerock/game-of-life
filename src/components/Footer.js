import React from 'react'
import styled from 'styled-components'

const FooterStyled = styled.div`
  background-color: #80cf7c;
  font-size: 1rem;
  padding: 5px;
  color: #2f2f2f;
  margin-top:0px;
  a {
    color: white;
    cursor: pointer;
    
    text-decoration: none;
  }
`

const I = styled.i`
  margin-left: 10px;

  cursor: pointer;
  color: #2f2f2f;
`

export class Footer extends React.Component {
  handleIconClick (link) {
    window.open('https://github.com/ReeceLangerock/game-of-life', '_blank')
  }
  render () {
    const currentYear = new Date().getFullYear()

    return (
      <FooterStyled className='Footer'>
        Â© {currentYear}{' '}
        <a href='http://reecelangerock.com' rel='noopener noreferrer' target='_blank'>
          Reece Langerock
        </a>
        <I className={`fa fa-github`} aria-hidden='false' onClick={() => this.handleIconClick()} />
      </FooterStyled>
    )
  }
}
export default Footer
