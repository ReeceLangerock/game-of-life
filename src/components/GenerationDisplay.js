import React from "react";
import styled, { injectGlobal } from "styled-components";

const StyledHeader = styled.div``;

export class GenerationDisplay extends React.Component {
  render() {
    return <div >{this.props.generationCount}</div>;
  }
}
export default GenerationDisplay;
