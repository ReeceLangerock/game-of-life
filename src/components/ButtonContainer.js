import React from "react";
import styled, { injectGlobal } from "styled-components";
import Button from './Button'

const StyledHeader = styled.div``;

export class ButtonContainer extends React.Component {
  render() {
    return (
      <div>
        <Button onClick={this.props.onRun}buttonText ='Run'/>
        <Button onClick={this.props.onPause} buttonText ='Pause'/>
        <Button onClick={this.props.clear}buttonText ='Clear'/>
      </div>
    );
  }
}
export default ButtonContainer;
