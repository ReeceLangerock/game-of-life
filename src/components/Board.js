import React from "react";
import styled, { injectGlobal } from "styled-components";
import Cell from "./Cell";
import gameOfLife from "../gameOfLife.js";
import ButtonContainer from "./ButtonContainer";

const StyledHeader = styled.div``;

const Container = styled.div`
  display: flex;
  flex-wrap: wrap;
  // width: ${props => 700 / props.size}px;
  width: 500px;
  border-right: 1px solid;
  border-bottom: 1px solid;
  margin: 0 auto;
`;

export class Board extends React.Component {
  constructor(props) {
    super(props);
    this.renderBoard = this.renderBoard.bind(this);
    this.handOfGod = this.handOfGod.bind(this);
    this.toggleTimer = this.toggleTimer.bind(this);
    this.resize = this.resize.bind(this);
    this.clear = this.clear.bind(this);
    this.startGameOfLife = this.startGameOfLife.bind(this);
    let boardSize = 50;
    console.log("const");
    const game = new gameOfLife(boardSize);

    this.state = {
      boardSize,
      gameBoard: new Array(boardSize),
      gameActive: false,
      game,
      generations: 0
    };
  }

  componentWillMount() {
    this.startGameOfLife();
  }

  timer() {
    let timer = setInterval(() => {
      // const game = new gameOfLife(this.state.boardSize);

      const updatedBoard = this.state.game.createNewGeneration(this.state.gameBoard);
      let generation = this.state.generations + 1;

      this.setState({
        gameBoard: updatedBoard,
        generations: generation
      });
    }, 50);
    this.setState({ timer });
  }

  startGameOfLife() {
    // console.log("gol", this.state.boardSize);
    // const game = new gameOfLife(this.state.boardSize);
    //  var t0 = performance.now();
    let board = this.state.game.initializeGame();
    //     var t1 = performance.now();
    // console.log("Call to initializeGame took " + (t1 - t0) + " milliseconds.");
    this.setState(
      {
        gameBoard: board
      }
      // console.log(this.state)
    );
  }

  handOfGod(index) {
    var t0 = performance.now();
    const board = this.state.gameBoard;
    board[index].alive = !board[index].alive;
    console.log(board[index]);
    this.setState(
      {
        gameBoard: board
      },
      () => {
        var t1 = performance.now();
        console.log("Call to initializeGame took " + (t1 - t0) + " milliseconds.");
      }
    );
  }

  toggleTimer(e) {
    console.log(e);

    if (!this.state.gameActive && e === "run") {
      this.timer();
      this.setState({
        gameActive: true
      });
    } else if (this.state.gameActive && e === "pause") {
      clearInterval(this.state.timer);
      this.setState({
        gameActive: false
      });
    }
  }

  resize() {
    this.setState(
      {
        boardSize: 40
      },
      () => {
        this.startGameOfLife();
      }
    );
  }

  clear() {
    let clearedBoard = this.state.game.clearBoard(this.state.gameBoard);
    this.setState({
      board: clearedBoard
    });
  }

  renderBoard() {
    return this.state.gameBoard.map(cell => {
      return <Cell ref={ref => (this.cellIndex = ref)} key={cell.index} status={cell.alive} size={this.state.boardSize} onClick={this.handOfGod.bind(this, cell.index)} />;
    });
  }

  render() {
    return (
      <div>
        {" "}
        <ButtonContainer onRun={this.toggleTimer.bind(this, "run")} onPause={this.toggleTimer.bind(this, "pause")} clear={this.clear} />
        <Container size={this.state.boardSize}>{this.renderBoard()}</Container>
        {this.state.generations}
      </div>
    );
  }
}
export default Board;
