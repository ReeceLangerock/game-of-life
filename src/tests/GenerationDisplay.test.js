import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, mount, find, render, simulate } from 'enzyme'
import enzymeSerializer from 'enzyme-to-json/serializer'
import { spy } from 'sinon'
import { GenerationDisplay } from './../components/GenerationDisplay'
import configureStore from 'redux-mock-store'

const initialState = {}
let mockStore = configureStore(initialState)
expect.addSnapshotSerializer(enzymeSerializer)
describe('GenerationDisplay', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<GenerationDisplay store={mockStore} />, div)
  })

  let component
  let props

  beforeEach(() => {
    props = {
      generationCount: 10
    }
    component = mount(<GenerationDisplay {...props} />)
  })

  it('displays correct generation count', () => {
    expect(component.text()).toEqual('10')
  })

  it('receives a generationCount prop', () => {
    expect(component.props().generationCount).toEqual(10)
  })


  it('matches snapshot', () => {
    const comp = shallow(<GenerationDisplay {...props} />)
    expect(comp).toMatchSnapshot()
  })
})
