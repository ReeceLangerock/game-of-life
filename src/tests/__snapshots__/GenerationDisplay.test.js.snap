// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerationDisplay matches snapshot 1`] = `
<div>
  10
</div>
`;
